core/include/cubos/core/net/tcp_stream.hpp

```diff
@@ -0,0 +1,95 @@
+/// @file
+/// @brief Class @ref cubos::core::net::TcpStream.
+/// @ingroup core-net
+
+#pragma once
+
+#include <cubos/core/api.hpp>
+#include <cubos/core/memory/stream.hpp>
+#include <cubos/core/net/address.hpp>
+#include <cubos/core/net/utils.hpp>
+
+namespace cubos::core::net
+{
+    /// @brief Stream implementation which writes to/reads from a TCP connection.
+    /// @ingroup core-net
+    class CUBOS_CORE_API TcpStream : public memory::Stream
+    {
+    public:
+        /// @brief Constructs an empty TCP stream, ready for connecting.
+        TcpStream();
+
+        /// @brief Deconstructs.
+        ~TcpStream() override;
+
+        /// @brief Forbid copy construction.
+        TcpStream(const TcpStream&) = delete;
+
+        /// @brief Forbid copy assignment.
+        TcpStream& operator=(const TcpStream&) = delete;
+
+        /// @brief Connects to a TCP listener, possibly on a remote machine.
+        /// @param address Destination address.
+        /// @param port Destination port.
+        /// @return Whether connecting was successful.
+        bool connect(const Address& address, uint16_t port);
+
+        /// @brief Replaces the inner socket with the given one.
+        /// @param socket Socket.
+        /// @return Whether connecting was successful.
+        void inner(InnerSocket socket);
+
+        /// @brief Disconnects inner socket.
+        void disconnect();
+
+        /// @brief Checks if inner socket is connected.
+        /// @return Whether the buffer holds a valid connection.
+        bool valid() const
+        {
+            return mSock != InnerInvalidSocket;
+        }
+
+        /// @brief Control blocking mode of inner socket.
+        /// @note A blocking socket does not return control until it has sent (or received) some or all data specified
+        /// for the operation.
+        /// @param blocking Whether to block or not.
+        /// @return Whether setting the block flag was sucessful.
+        bool setBlocking(bool blocking);
+
+        /// @brief Reads data from the stream with the exact specified size.
+        /// @param data Buffer to read data into.
+        /// @param size Size of the buffer.
+        /// @note This function will only return after buffer is completed or an error occurs.
+        /// @return Whether reading was sucessful.
+        bool readExact(void* data, std::size_t size);
+
+        /// @brief Writes data to the stream with the exact specified size.
+        /// @param data Buffer to write data from.
+        /// @param size Size of the buffer.
+        /// @note This function will only return after whole buffer is written or an error occurs.
+        /// @return Whether writing was sucessful.
+        bool writeExact(const void* data, std::size_t size);
```

Comment:
These utility methods should be part of the base class, no? Also, a note should be added to the base Stream::read/write method warning that the data may be written partially, and a recommendation to use the `*Exact` methods in case that's a problem.